### ISSUE TITLE
Use `#[serde(deny_unknown_fields)]` on non-changeble types

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#2224]
Types like `Transaction` (and its components) will never change fields and shouldn't parse unknown fields. We should deny unknown fields (and if we stop using derive write equivalent code that denies them).

**Comment by apoelstra** [comment]
This will break if anybody is parsing data from the Bitcoin Core RPC directly into `Transaction` types.

**Comment by Kixunil** [comment]
Oh, didn't realize Core adds more fields. That's annoying. But also I'm starting kinda dislike the whole thing. (De)serialization cannot be tweaked with `serde` so people have to write their own modules, which is annoying. And blindly using whatever we have is not great for new protocols. Maybe we should provide modules for alternate forms.

**Comment by apoelstra** [comment]
>Oh, didn't realize Core adds more fields. That's annoying. 

Yeah, it used to cause us a lot of grief back before we used serde/serde_json, which (among other things) ignored unknown fields by default.

>Maybe we should provide modules for alternate forms.

Well, Core's "protocol" is served by the bitcoin-core-rpc crate; the consensus byte encoding by this one; PSBT by this crate (though we don't do a very good job). Other protocols are probably out of scope for this project.

I guess you mean, we should have some stability guarantee for the serde encoding of our types, which I consider to be a kind of pickling and not something that should be used in real protocols. And maintaining that stability shouldn't impose restrictions on us, by e.g. exposing the private fields of our types. Agreed on both counts. But I also don't want to complicate our API by making the user pick a "real protocol" before being able to use serde.

**Comment by Kixunil** [comment]
OK, then I think we should document that the serialization is specifically mean to be compatible with Core and make it a stability guarantee. For anything that doesn't have a known used serialization I would rather avoid providing any.

I meant that accepting unknown fields is horrible for debugging. For instance a few months ago I discovered a subtle bug in my ode because of malformed toml only after implementing stricter deserialization that rejects unknown fields. However if this affects Core integration I am willing to let it be this way provided it's documented. And recommend people to use (hex) consensus serialization for new protocols.

> exposing the private fields of our types

Types with private fields should never be (de)serialized unless those fields are caching data within that type. (Like in `script::Builder`).

**Comment by apoelstra** [comment]
Ok, yeah, agreed on all counts. Both explicitly documenting "the serde impls are there for use with JSON-RPC interfaces, in particular Core, and will silently ignore unknown fields" and "serious protocols should use the hex/binary consensus encoding"

**Comment by tcharding** [comment]
Has Core's JSON-RPC interface ever change the serialization format of types, because then we would be tied to certain versions of Core and we'd probably need to test and document that?

**Comment by apoelstra** [comment]
@tcharding in a couple instances yes, but Core tries very hard not to do this (and it has become less frequent over time).

AFAIK they have never changed the format of the primitive types (hashes, txes, blocks, etc) except in extraneous fields (e.g. `addresses`) which we don't parse anyway.

**Comment by Kixunil** [comment]
Closing in favor of #2372 for clarity.
